app/scripts/utils/throttle-and-debounce.js

```diff
@@ -0,0 +1,52 @@
+const throttleAndDebounce = (func, interval, finalWait) => {
+  let timeout;
+  let blockedCalls = 0;
+
+  const reset = () => {
+    timeout = null;
+  };
+
+  const debounced = (...args) => {
+    const later = () => {
+      // Since we throttle and debounce we should check whether there were
+      // actually multiple attempts to call this function after the most recent
+      // throttled call. If there were no more calls we don't have to call
+      // the function again.
+      if (blockedCalls > 0) {
+        func(...args);
+        blockedCalls = 0;
+      }
+    };
+
+    clearTimeout(timeout);
+    timeout = setTimeout(later, finalWait);
+  };
+
+  debounced.cancel = () => {
+    clearTimeout(timeout);
+    reset();
+  };
+
+  debounced.immediate = (...args) => {
+    func(...args);
+  };
+
+  let wait = false;
+  const throttled = (request, ...args) => {
```

Comment:
It would be nice to have a short explanation of why we're using throttling and how it works together with debouncing. I know you wrote this in the PR but if someone is looking at the code they won't have access to the PR.

Reply:
Good point. I wanted to do that but somehow forgot. Will add an explanation.

Reply:
Done!